### DISCUSSION TITLE
Pyinstaller unable to build with jsons not working properly

### DISCUSSION THREAD

**Discussion by DasWildeMaus** [pyinstaller/pyinstaller#8879] [Help]
Hey guys, i have a problem, that I can't build a working .exe by using pyinstaller and none of the other problems got me close to a solution with my issue

the cmd command was:

`pyinstaller --onedir --windowed --icon=Roulette-Bot.ico --add-data "src/*.json;." --add-data "src/assets;assets" --add-data "src/data;data" --add-data "src/tesseract;tesseract" src\gui.py`

but it seems to be unable to work with the jsons properly. If I open the .exe the jsons can't be found

Structure:
Roulette-Bot-Code/
├── .idea/
├── .venv/
├── build/
├── config/
├── dist/
├── logs/
│ └── app.log
├── src/
│ ├── assets/
│ ├── data/
│ ├── tesseract/
│ ├── betting_logic.py
│ ├── calibration.json
│ ├── [calibration.py](http://calibration.py/)
│ ├── [capture.py](http://capture.py/)
│ ├── chip_calibration.json
│ ├── [gui.py](http://gui.py/)
│ ├── number_recognition_calibration.json
│ ├── progression_table.py
│ ├── [recognition.py](http://recognition.py/)
│ ├── series_calibration.json
│ └── settings.json
├── Tesseract OCR/
├── tests/ │
├── test_calibration.py │
├── test_capture.py
│ └── test_recognition.py
├── .gitignore
├── gui.spec
├── [README.md](http://readme.md/)
├── README - [Kopie.md](http://kopie.md/)
└── Roulette-Bot.ico

I tried different --add-data commands but nothing works. Always [name].json can't be found for any of the jsons in the src folder

And there are two other jsons in the config folder, which are working fine. But they are rather unimportant for quick changes so I'd like to have the other jsons directly in the source folder as they are atm.

Thank you

**Comment by rokm** [comment]
> I tried different --add-data commands but nothing works.

What exactly does "nothing works" mean - that you cannot see the .json files in `dist/<name>/_internal` if you look into that directory (e.g., with file browser), or that your code cannot find them at runtime (for example, because you are using relative paths instead of absolute paths as described in [the docs](https://pyinstaller.org/en/stable/runtime-information.html#using-file))?

**Reply by DasWildeMaus** [reply]
![grafik](https://github.com/user-attachments/assets/c02da3a6-a242-42ab-9dbf-d0e82a54fe7e)
that's the error. 
And the _internal folder doesn't contain any of the jsons.
I'm not sure though about the paths as it is my first project that I tried to export with an exe to open.
Currently the line in the error points to `with open("series_calibration.json", "r") as file:`

At the top of the file I have 
```
SETTINGS_FILE = "settings.json"
OUTPUT_PATH = Path(__file__).parent
ASSETS_PATH = OUTPUT_PATH / Path(r"assets/frame0")
program_running = False
print("ASSETS_PATH:", ASSETS_PATH)


def relative_to_assets(path: str) -> Path:
    return ASSETS_PATH / Path(path)

print(relative_to_assets("button_2_stopped.png"))
print(relative_to_assets("button_2_running.png"))
```
that's done with a bit of chatGPT trying to fix the error. By now I don't even know anymore how I had it before.

How would the paths need to be exactly? Any way it is in the docs?

**Reply by rokm** [reply]
>  Currently the line in the error points to `with open("series_calibration.json", "r") as file:`

This assumes that `series_calibration.json` is in the current working directory. Which is seldom the case with frozen application. So even if the .json files were properly collected, you would be getting this error.

>  And the _internal folder doesn't contain any of the jsons.

What are the contents, then? I find it a bit odd that `assets` directory was collected but .json files were not... (plus, `--add-data "src/*.json;."` seems to work as expected with test directory layout that I have made).

> How would the paths need to be exactly? Any way it is in the docs?

Either anchor it to parent of `__file__` attribute (which should work with both unfrozen script and frozen application), or to `sys._MEIPASS` (which is specific to frozen application).

https://pyinstaller.org/en/stable/runtime-information.html#using-file

**Reply by DasWildeMaus** [reply]
I copied the project to another folder and did it again and now the jsons are in the _internals:

![grafik](https://github.com/user-attachments/assets/62296b92-22c1-4e82-bb0d-a0f54205d075)
and in the _internal is:
![grafik](https://github.com/user-attachments/assets/439e7e53-623b-4dc3-ab68-136ce58a4b43)
[a lot of libs in between]
![grafik](https://github.com/user-attachments/assets/6e69cff3-c746-4fd8-bd42-5bf3973264ed)

> This assumes that series_calibration.json is in the current working directory. Which is seldom the case with frozen application. So even if the .json files were properly collected, you would be getting this error.

What exactly does this line mean and how would the jsons then need to be implemented?
As that seems to be really the only error and quite a rookie mistake

**Reply by rokm** [reply]
> What exactly does this line mean

Exactly what it says. 

You are trying to open "series_calibration.json", which, because it is a relative path, assumes that the file is in the current working directory. Even for your unfrozen `gui.py` script, that works only if you run it from the `src` directory (i.e., with `src` being your current working directory). If you run it from anywhere else, you will end up with the same file-not-found error. And similarly, in frozen application, the current working directory is typically the directory that contains the executable, not the `_internal` directory (which contains your data).

So as per docs that I've linked earlier, you should anchor paths to your data (in this case .json) files to the module's (or entry-point script's) `__file__` attribute:

```
json_file = os.path.join(os.path.dirname(__file__), "series_calibration.json")
with open(json_file, "r") as file:
   ...
```

**Reply by DasWildeMaus** [reply]
alright thank you I'll do that.
Does that mean if I just had put the jsons in the /src folder, it wouldn't have been a problem in the first place?

**Reply by rokm** [reply]
>  Does that mean if I just had put the jsons in the /src folder, it wouldn't have been a problem in the first place?

No, the issue is not location of those files per-se, but how you are referring to them in your code.

For example, open command prompt, activate your python environment, then move to the top-level `Roulette-Bot-Code` directory, and try to run your program from there - `python Roulette-Bot-Code\src\gui.py`. If you use relative paths (e.g., `with open("series_calibration.json", "r") as file:˙) you will also get file-not-found errors because files are not present in your current working directory.

**Reply by DasWildeMaus** [reply]
ok I fixed all the paths now.

When typing into the cmd:
`pyinstaller --onedir --windowed --icon="Roulette-Bot.ico" --add-data "src/calibration.json;src" --add-data "src/chip_calibration.json;src" --add-data "src/number_recognition_calibration.json;src" --add-data "src/series_calibration.json;src" --add-data "src/settings.json;src" --add-data "src/assets;src/assets" --add-data "src/data;src/data" --add-data "src/tesseract;src/tesseract" src/gui.py`

it says

`Unable to find '[pathToProj]\\Roulette-Bot-Code\\src\\assets' when adding binary and data files.`
with pathToProj being the correct folder structure to the project "Roulette-Bot-Code"

so I don't even get to build the project
------
![grafik](https://github.com/user-attachments/assets/59309d79-d83a-4dc6-8ba1-603b4e1ff5ba)

**Reply by rokm** [reply]
That error is raised here

https://github.com/pyinstaller/pyinstaller/blob/ec99a08ab0ed7eb22770aa3b10cc70f7ea3ff59d/PyInstaller/building/utils.py#L486-L495

when the source path does not point to the file and the resulting `glob.glob()` call comes back empty. But your screenshot indicates that the directory does seem to exist - does `pathToProj` perhaps contain characters that would mess up the `glob.glob()` call (for example, square brackets)?

**Reply by DasWildeMaus** [reply]
oh yeah indeed it has square brackets damn. 
Copied the whole proj to desktop to run the pyinstaller but now the jsons are missing again💀 
I ran the exact same cmd as above
`pyinstaller --onedir --windowed --icon="Roulette-Bot.ico" --add-data "src/calibration.json;src" --add-data "src/chip_calibration.json;src" --add-data "src/number_recognition_calibration.json;src" --add-data "src/series_calibration.json;src" --add-data "src/settings.json;src" --add-data "src/assets;src/assets" --add-data "src/data;src/data" --add-data "src/tesseract;src/tesseract" src/gui.py`
![grafik](https://github.com/user-attachments/assets/909a72dd-f6ae-4f3e-8c7b-727c5295139d)
Is that some pathing that I need to adjust when switching the directory? As before on the "wrong" pathing, the jsons were at least copied into _internals?
So basically I moved the proj folder from `D:[pathToFile]/Roulette-Bot-Code` directly onto the desktop

**Reply by rokm** [reply]
Well, you are not collecting .json files into top-level application directory anymore, but into `src` sub-directory; so you can find them in `_internal\src`. 

You probably want to omit `src` in the destination arguments of your `--add-data` arguments, i.e., `--add-data "src/calibration.json;." --add-data "src/chip_calibration.json;." --add-data "src/number_recognition_calibration.json;." --add-data "src/series_calibration.json;." --add-data "src/settings.json;." --add-data "src/assets;assets" --add-data "src/data;data" --add-data "src/tesseract;tesseract"`

**Reply by DasWildeMaus** [reply]
oh my god that worked thank you so much for your assistance, I learned a lot
